universal-application-tool-0.0.1/app/controllers/admin/QuestionController.java

```diff
@@ -0,0 +1,113 @@
+    this.listView = checkNotNull(listView);
+    this.editView = checkNotNull(editView);
+    this.formFactory = checkNotNull(formFactory);
+  }
+
+  public CompletionStage<Result> create(Request request) {
+    Form<QuestionForm> form = formFactory.form(QuestionForm.class);
+    QuestionForm questionForm = form.bindFromRequest(request).get();
+    return service
+        .getReadOnlyQuestionService()
+        .thenApplyAsync(
+            readOnlyService -> {
+              try {
+                QuestionDefinition definition =
+                    questionForm
+                        .getBuilder()
+                        .setId(readOnlyService.getNextId())
+                        .setVersion(1L)
+                        .build();
+                service.create(definition);
+              } catch (UnsupportedQuestionTypeException e) {
+                // I'm not sure why this would happen here, so we'll just log and redirect.
+                LOG.info(e.toString());
+              }
+              return redirect("/admin/questions");
+            });
+  }
+
+  public CompletionStage<Result> edit(Request request, String path) {
+    return service
+        .getReadOnlyQuestionService()
+        .thenApplyAsync(
+            readOnlyService -> {
+              Optional<QuestionDefinition> definition = Optional.empty();
+              try {
+                definition = Optional.of(readOnlyService.getQuestionDefinition(path));
+              } catch (InvalidPathException e) { // If the path doesn't exist, redirect to create.
+                LOG.info(e.toString());
+              }
+              if (definition.isPresent()) {
+                return ok(editView.render(request, definition));
+              } else {
+                return redirect("/admin/questions/new");
+              }
+            });
+  }
+
+  public Result newOne(Request request) {
+    return ok(editView.render(request, Optional.empty()));
+  }
+
+  public CompletionStage<Result> index(String renderAs) {
+    return service
+        .getReadOnlyQuestionService()
+        .thenApplyAsync(
+            readOnlyService -> {
+              return ok(listView.render(readOnlyService.getAllQuestions(), renderAs));
+            });
+  }
+
+  public CompletionStage<Result> update(Request request, Long id) {
+    Form<QuestionForm> form = formFactory.form(QuestionForm.class);
+    QuestionForm questionForm = form.bindFromRequest(request).get();
+    try {
+      QuestionDefinition definition = questionForm.getBuilder().setId(0L).setVersion(1L).build();
+      service.update(definition);
+    } catch (UnsupportedQuestionTypeException e) {
+      // I'm not sure why this would happen here, so we'll just log and redirect.
+      LOG.info(e.toString());
+    } catch (UnsupportedOperationException e) {
+      // This is expected for now until we implement update on QuestionService.
+    }
+    return CompletableFuture.completedFuture(redirect("/admin/questions"));
```

Comment:
Suggestion (blocking): use reverse router for path

Reply:
Done